ci-operator/step-registry/openshift/e2e/azure/csi/secrets-store/vault/openshift-e2e-azure-csi-secrets-store-vault-workflow.yaml

```diff
@@ -0,0 +1,30 @@
+workflow:
+  as: openshift-e2e-azure-csi-secrets-store-vault
+  steps:
+    pre:
+    - chain: ipi-azure-pre
+    - ref: optional-operators-subscribe
+    - ref: storage-create-csi-secrets-store
+    - ref: storage-conf-wait-for-csi-driver
+    - ref: storage-obj-save
+    test:
+    - as: test-e2e-vault
+      cli: latest
+      commands: |
+        # upstream bats use kubectl but the cli image only installs oc
+        alias kubectl=oc
+        make e2e-vault
```

Comment:
This is a good idea, but it didn't work in my rehearse job... did I do something different?

Reply:
Hi @dobsonj , it seems the previous job failed not as expected(seems still use the old codes), the latest on is the actual result, the `oc create ns $(oc project -q)` not work well.
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_release/51365/rehearse-51365-pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault/1785116980139790336/artifacts/operator-e2e-vault/test-e2e-vault/build-log.txt

```console
error: no project has been set
error: exactly one NAME is required, got 0
See 'oc create namespace -h' for help and examples
```

Reply:
Apparently I was looking at stale test results :) I pushed a new commit just now, we'll see.

Reply:
I guess `oc project e2e-azure-csi-secrets-store-vault` maybe not work since seems the kubeconfig is in the shared_dir with readonly mode, but the command switch the context want to update the kubeconfig file. If we want to workround for the issue, maybe we could write the kubeconfig to $HOME some dir and export the KUBECONFIG again. Anyway we could wait for the result and try later, that's my suppose and I remember seems I meet some similar issue before.

Reply:
The `oc project e2e-azure-csi-secrets-store-vault` worked, seems the kubeconig file not readonly, but seem still have too many errors, `# /tmp/bats-run-152-eQbgER/bats.192.src: line 23: helm: command not found`, we still needs to install helm in the test image dockerfile. I could try to temporarily download in test phase to test what else we needed.

Reply:
Even if I temporarily  install the helm in test phase, it seems the test codes still has bug, seems the similar issue which @jsafrane mentioned in https://github.com/openshift/secrets-store-csi-driver/pull/16#discussion_r1567554801 . @ropatil010 Do you happen to know if there's something we missed or it needs to set some other env vars?

```console
sh-5.1$ oc get po -n vault
NAME                       READY   STATUS             RESTARTS   AGE
vault-0                    0/1     ImagePullBackOff   0          9m49s
vault-csi-provider-qhz2p   1/2     ImagePullBackOff   0          9m42s
vault-csi-provider-ws8mv   1/2     ImagePullBackOff   0          9m47s
vault-csi-provider-zsjhw   1/2     ImagePullBackOff   0          9m42s

$ sh-5.1$ oc describe po vault-0 -n vault
Name:             vault-0
Namespace:        vault
Priority:         0
Service Account:  vault
Node:             ci-op-xkw6777t-2cad1-qjd58-worker-eastus22-gg79w/10.0.128.6
Start Time:       Tue, 30 Apr 2024 09:25:46 +0000
Labels:           app.kubernetes.io/instance=vault
                  app.kubernetes.io/name=vault
                  apps.kubernetes.io/pod-index=0
                  component=server
                  controller-revision-hash=vault-945764cc6
                  helm.sh/chart=vault-0.28.0
                  statefulset.kubernetes.io/pod-name=vault-0
Annotations:      k8s.ovn.org/pod-networks:
                    {"default":{"ip_addresses":["10.130.2.27/23"],"mac_address":"0a:58:0a:82:02:1b","gateway_ips":["10.130.2.1"],"routes":[{"dest":"10.128.0.0...
                  k8s.v1.cni.cncf.io/network-status:
                    [{
                        "name": "ovn-kubernetes",
                        "interface": "eth0",
                        "ips": [
                            "10.130.2.27"
                        ],
                        "mac": "0a:58:0a:82:02:1b",
                        "default": true,
                        "dns": {}
                    }]
                  openshift.io/scc: restricted-v2
                  seccomp.security.alpha.kubernetes.io/pod: runtime/default
Status:           Pending
SeccompProfile:   RuntimeDefault
IP:               10.130.2.27
IPs:
  IP:           10.130.2.27
Controlled By:  StatefulSet/vault
Containers:
  vault:
    Container ID:  
    Image:         hashicorp/vault:1.16.1
    Image ID:      
    Ports:         8200/TCP, 8201/TCP, 8202/TCP
    Host Ports:    0/TCP, 0/TCP, 0/TCP
    Command:
      /bin/sh
      -ec
    Args:
      /usr/local/bin/docker-entrypoint.sh vault server -dev 
      
    State:          Waiting
      Reason:       ImagePullBackOff
    Ready:          False
    Restart Count:  0
    Readiness:      exec [/bin/sh -ec vault status -tls-skip-verify] delay=5s timeout=3s period=5s #success=1 #failure=2
    Environment:
      HOST_IP:                    (v1:status.hostIP)
      POD_IP:                     (v1:status.podIP)
      VAULT_K8S_POD_NAME:        vault-0 (v1:metadata.name)
      VAULT_K8S_NAMESPACE:       vault (v1:metadata.namespace)
      VAULT_ADDR:                http://127.0.0.1:8200
      VAULT_API_ADDR:            http://$(POD_IP):8200
      SKIP_CHOWN:                true
      SKIP_SETCAP:               true
      HOSTNAME:                  vault-0 (v1:metadata.name)
      VAULT_CLUSTER_ADDR:        https://$(HOSTNAME).vault-internal:8201
      HOME:                      /home/vault
      VAULT_DEV_ROOT_TOKEN_ID:   root
      VAULT_DEV_LISTEN_ADDRESS:  [::]:8200
    Mounts:
      /home/vault from home (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-24ns8 (ro)
Conditions:
  Type                        Status
  PodReadyToStartContainers   True 
  Initialized                 True 
  Ready                       False 
  ContainersReady             False 
  PodScheduled                True 
Volumes:
  home:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  kube-api-access-24ns8:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
    ConfigMapName:           openshift-service-ca.crt
    ConfigMapOptional:       <nil>
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason          Age                   From               Message
  ----     ------          ----                  ----               -------
  Normal   Scheduled       13m                   default-scheduler  Successfully assigned vault/vault-0 to ci-op-xkw6777t-2cad1-qjd58-worker-eastus22-gg79w
  Normal   AddedInterface  13m                   multus             Add eth0 [10.130.2.27/23] from ovn-kubernetes
  Normal   Pulling         12m (x4 over 13m)     kubelet            Pulling image "hashicorp/vault:1.16.1"
  Warning  Failed          12m (x4 over 13m)     kubelet            Failed to pull image "hashicorp/vault:1.16.1": reading manifest 1.16.1 in registry.connect.redhat.com/hashicorp/vault: name unknown: Image not found
  Warning  Failed          12m (x4 over 13m)     kubelet            Error: ErrImagePull
  Warning  Failed          11m (x6 over 13m)     kubelet            Error: ImagePullBackOff
  Normal   BackOff         3m43s (x40 over 13m)  kubelet            Back-off pulling image "hashicorp/vault:1.16.1"
```

Reply:
I guess we still needs 
```console
helm install vault hashicorp/vault \
    --set "global.openshift=true" \
    --set "server.dev.enabled=true" \
    --set "server.image.repository=docker.io/hashicorp/vault" \
    --set "injector.image.repository=docker.io/hashicorp/vault-k8s"

```
which also mentioned in https://developer.hashicorp.com/vault/tutorials/kubernetes/kubernetes-openshift#install-the-vault-helm-chart

Reply:
Not sure, but i still followed my older steps on 4.15 cluster now and is working fine.  
More clear steps here: https://issues.redhat.com/browse/OSDOCS-10136 

rohitpatil@ropatil-mac Downloads % oc get pods -n vault 
NAME                       READY   STATUS    RESTARTS   AGE
vault-0                    1/1     Running   0          42s
vault-csi-provider-58vzj   1/2     Running   0          5s
vault-csi-provider-bmnx2   1/2     Running   0          5s
vault-csi-provider-kz4fr   1/2     Running   0          5s

Reply:
4.15 describe pod
 Normal  Pulled          29m   kubelet            Container image "hashicorp/vault-csi-provider:1.4.2" already present on machine
  Normal  Created         29m   kubelet            Created container vault-csi-provider
  Normal  Started         29m   kubelet            Started container vault-csi-provider
  Normal  Pulled          29m   kubelet            Container image "hashicorp/vault:1.16.1" already present on machine
  Normal  Created         29m   kubelet            Created container vault-agent
  Normal  Started         29m   kubelet            Started container vault-agent

Reply:
@Phaow @ropatil010 I opened https://github.com/openshift/secrets-store-csi-driver/pull/18 which should fix these issues (missing tools, and pulling from the wrong registry). I'll update this PR and retest again after that merges.

One thing I found out while testing it manually is that `make e2e-vault` assumes that the tests are run from the `default` namespace. If we execute it from a different namespace, some of the test pods fail to start with:
`  Warning  FailedMount  2m13s                kubelet            MountVolume.SetUp failed for volume "secrets-store-rotation" : rpc error: code = Unknown desc = failed to mount secrets store objects for pod e2e-azure-csi-secrets-store-vault/secrets-store-rotation, err: rpc error: code = Unknown desc = error making mount request: couldn't read secret "foo": failed to login: Error making API request.`

Reply:
This time the [tests mostly passed](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_release/51365/rehearse-51365-pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault/1785428695490498560/artifacts/operator-e2e-vault/test-e2e-vault/build-log.txt), except:
```
# daemonset.apps/vault-csi-provider patched
# pod/vault-0 condition met
# pod/vault-csi-provider-hm7r7 condition met
# Error from server (NotFound): pods "vault-csi-provider-7dlfw" not found
# Error from server (NotFound): pods "vault-csi-provider-k5f78" not found
```
So this [wait condition](https://github.com/openshift/secrets-store-csi-driver/blob/b3a96ae272638f029c8566f07ebfffd67e63c426/test/bats/vault.bats#L39) fails because the pods restarted, but the rest of the tests passed after that. I think this is just a bug in the test, there is a 10 second sleep before that, which may not be enough time for the new pods to be created.

At least the rehearsal is working now, I think @ropatil010 can pull these changes back into his PR now.